File: Baekjoon/Tree/2042.cpp
#include <iostream>
using namespace std;

int n, m, k;
long long tree[1024 * 1024 * 2];
int tmpN = 0;

// 1. tree의 배열 크기 tmpN 구하기
// 1-1 tmpN = leaf 노드의 개수
// 1-2 tmpN * 2 == 실제 배열의 크기
// 2. tree 값 초기화
// 2-1. inpur leafnode
// 2-2 Internal node 계산
// 3. update
// 4. 구간 합 구하기

void Update(int a, long long b)
{
	// a 번째 값 -> tmpN
	a = a + tmpN - 1;
	tree[a] = b;
	while (a > 1)
	{
		a = a >> 1;
		tree[a] = tree[a * 2] + tree[a * 2 + 1];
	}
}

long long getSum(int a, int b)
{
	a = a + tmpN - 1;
	b = b + tmpN - 1;
	long long sum = 0;

	while (a <= b)
	{
		// a -> right child? ->라면 오른쪽 부모로 올라감
		if ((a & 1) == 1) sum += tree[a];
		// b -> left child? -> 라면 왼쪽 부모로 올라감
		if ((b & 1) == 0) sum += tree[b];
		a = (a + 1) >> 1;
		b = (b - 1) >> 1;
	}
	return (sum);
}

int	main(void)
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	cin >> n >> m >> k;
	for (tmpN = 1; tmpN < n; tmpN *= 2);
	for (int i = 0; i < n; i++)
	{
		cin >> tree[tmpN + i];
	}
	for (int i = tmpN + n; i < tmpN * 2; i++)
		tree[i] = 0;
	for (int i = tmpN - 1; i >= 1; i--)
	{
		tree[i] = tree[2 * i] + tree[2 * i + 1];
	}

	long long a, b, c;
	for (int i = 0; i < m + k; i++)
	{
		cin >> a >> b >> c;
		if (a == 1)
		{
			Update(b, c);
		}
		else
		{
			cout << getSum(b, c) << "\n";
		}
	}
	return (0);
}

File: Baekjoon/Java/Tree/2042.java
package com.company;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static long[] tree;
    static long[] num;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader((System.in)));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int level = 1;

        while (level < n) {
            level *= 2;
        }
        tree = new long[level * 2];
        num = new long[level];

        for (int i=0; i<n; i++) {
            num[i] = Long.parseLong(br.readLine());
        }

        init(n, level);

        for (int i=0; i<m+k; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            long c = Long.parseLong(st.nextToken());
            if (a == 1) {
                update(1, 1, level, b, c - num[b-1]);
                num[b-1] = c;
            }
            else {
                System.out.println(sum(1, 1, level, b, (int)c));
            }
        }
    }

    static void init(int n, int level) {
        for (int i=0; i<n; i++) {
            tree[i + level] = num[i];
        }
        for (int i=level-1; i>0; i--) {
            tree[i] = tree[i*2] + tree[i*2+1];
        }
    }

    static void update(int node, int start, int end, int index, long diff) {
        if (index < start || index > end) return ;
        tree[node] += diff;
        if (start == end) return ;
        update(node*2, start, (start+end)/2, index, diff);
        update(node*2+1, (start+end)/2+1, end, index, diff);
    }

    static long sum(int node, int start, int end, int left, int right) {
        if (left > end || right < start) return (0);
        if (left <= start && end <= right) return tree[node];
        return sum(node*2, start, (start+end)/2, left, right) +
                sum(node*2+1, (start+end)/2+1, end, left, right);
    }
}
